workflow/controller/workflowpod.go

```diff
@@ -727,51 +728,74 @@ func addVolumeReferences(pod *apiv1.Pod, vols []apiv1.Volume, tmpl *wfv1.Templat
 		return nil
 	}
 
-	addVolumeRef := func(volMounts []apiv1.VolumeMount) error {
-		for _, volMnt := range volMounts {
-			vol := getVolByName(volMnt.Name)
-			if vol == nil {
-				return errors.Errorf(errors.CodeBadRequest, "volume '%s' not found in workflow spec", volMnt.Name)
-			}
-			found := false
-			for _, v := range pod.Spec.Volumes {
-				if v.Name == vol.Name {
-					found = true
-					break
-				}
+	for _, volMnt := range volMounts {
+		vol := getVolByName(volMnt.Name, vols, tmpl, pvcs)
+		if vol == nil {
+			return errors.Errorf(errors.CodeBadRequest, "volume '%s' not found in workflow spec", volMnt.Name)
+		}
+		found := false
+		for _, v := range pod.Spec.Volumes {
+			if v.Name == vol.Name {
+				found = true
+				break
 			}
-			if !found {
-				if pod.Spec.Volumes == nil {
-					pod.Spec.Volumes = make([]apiv1.Volume, 0)
-				}
-				pod.Spec.Volumes = append(pod.Spec.Volumes, *vol)
+		}
+		if !found {
+			if pod.Spec.Volumes == nil {
+				pod.Spec.Volumes = make([]apiv1.Volume, 0)
 			}
+			pod.Spec.Volumes = append(pod.Spec.Volumes, *vol)
 		}
+	}
+	return nil
+}
+
+// addPodPatchVolRef adds any volumeMounts that PodSpecPatch referencing, to the pod.spec.volumes
+func addPodPatchVolRef(patchPodSpec apiv1.PodSpec, pod *apiv1.Pod, vols []apiv1.Volume, tmpl *wfv1.Template, pvcs []apiv1.Volume) error {
```

Comment:
many of these parameters are in woc. 

Change `func addPodPatchVolRef(` to `func (woc *...) addPodPatchVolRef(` and make func simpler